### PR TITLE
Power up pre-commit

### PR DESCRIPTION
**Pull request by basnijholt**
This adds some whitespace checks, `isort`, `pyupgrade`, and some other code checks from the `pre-commit-hooks` repo.

Note that `pyupgrade` doesn't really change much here because I used this already before.

A summary of what the hooks do:
* `pyupgrade` - A tool (and pre-commit hook) to automatically upgrade syntax for newer versions of the language.
* `end-of-file-fixer` - Makes sure files end in a newline and only a newline.
* `check-docstring-first` - Checks for a common error of placing code before the docstring.
* `check-yaml` - Attempts to load all yaml files to verify syntax.
* `debug-statements` - Check for debugger imports and py37+ breakpoint() calls in python source.
* `check-ast` - Simply check whether files parse as valid python.
* `seed-isort-config` - Statically populate the `known_third_party` `isort` setting because `isort` sucks at figuring out which third party libs are used. (from author of `pre-commit`)

I have also added this summary to the first commit.

### CODE REVIEW

File: .isort.cfg [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[settings]
+known_third_party = PIL,atomicwrites,holoviews,ipykernel,matplotlib,nbconvert,numpy,pytest,scipy,setuptools,skopt,sortedcollections,sortedcontainers,zmq
```
Comment (jbweston) [inline .isort.cfg]: what does this do?

Will this mean that if we ever add another library to the dependencies we'll need to add it here too?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -1,16 +1,43 @@
...
+-   repo: https://github.com/ambv/black
+    rev: 19.10b0
+    hooks:
+    - id: black
+      language_version: python3.7
+-   repo: https://github.com/asottile/pyupgrade
+    rev: v1.25.2
+    hooks:
+    -   id: pyupgrade
+        args: ['--py36-plus']
```
Comment (jbweston) [inline .pre-commit-config.yaml]: Can you either doctor the commit message or add a comment about what each of these things does? e.g. I don't know what `pyupgrade` is about

File: .isort.cfg (hunk 1 above)
Comment (basnijholt) [reply .isort.cfg]: `isort` isn't very good at figuring out what libraries are third party libs. The pre-commit hook `seed-isort-config` automatically maintains this file.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (basnijholt) [reply .pre-commit-config.yaml]: See my updated comment [here](https://github.com/python-adaptive/adaptive/pull/237#issue-351901333)

Comment (basnijholt) [reply .pre-commit-config.yaml]: and https://github.com/python-adaptive/adaptive/pull/237/commits/0a5cbe1d40a6b11a99c1b5262b411fc43d936db4

**Comment by basnijholt** [comment]
isort is [able to deduce the requirements from `requirements.txt`](https://github.com/timothycrosley/isort/blob/c23f76f8ed020455a78d9480ef9e52f476cf601b/isort/finders.py#L306-L317). We could also define that and get rid of the `.isort.cfg` file.

File: .isort.cfg (hunk 1 above)
Comment (basnijholt) [reply .isort.cfg]: `isort` is [able to deduce the requirements from `requirements.txt`](https://github.com/timothycrosley/isort/blob/c23f76f8ed020455a78d9480ef9e52f476cf601b/isort/finders.py#L306-L317). We could also define that and get rid of the `.isort.cfg` file.

Comment (jbweston) [reply .isort.cfg]: I was under the impression that `requirements.txt` was really for deploying apps, and not really for library development.

How does `seed-isort-config` work?

Comment (basnijholt) [reply .isort.cfg]: It [reads the source files](https://github.com/asottile/seed-isort-config/blob/master/seed_isort_config.py) IIUC.

Comment (jbweston) [reply .isort.cfg]: If `seed-isort-config` maintains this file, then we shouldn't check it in to git, right? It will just be autogenerated by `pre-commit`

Comment (basnijholt) [reply .isort.cfg]: It's not possible to not commit the file:
```bash
(base) basnijholt-macbook  ➜  adaptive git:(power-up-pre-commit) ✗  rm .isort.cfg
(base) basnijholt-macbook  ➜  adaptive git:(power-up-pre-commit) ✗  pre-commit run --all
Trim Trailing Whitespace.................................................Passed
Fix End of Files.........................................................Passed
Check docstring is first.................................................Passed
Check Yaml...............................................................Passed
Debug Statements (Python)................................................Passed
Check python ast.........................................................Passed
flake8...................................................................Passed
black....................................................................Passed
pyupgrade................................................................Passed
isort....................................................................Passed
seed isort known_third_party.............................................Failed
hookid: seed-isort-config

Files were modified by this hook. Additional output:

Creating an .isort.cfg with a known_third_party setting. Feel free to move the setting to a different config file in one of .editorconfig, .isort.cfg, setup.cfg, tox.ini, pyproject.toml.

This setting should be committed.
```

Comment (jbweston) [reply .isort.cfg]: never mind; I just tried it locally and it complains that `.isort.cfg` should be committed.

**Pull request merged**